### PR TITLE
CDRIVER-4372 Add FLE 2 API to explicit encryption

### PR DESCRIPTION
**Pull request by kevinAlbs**
# Summary

- Support new encryption algorithms "Indexed" and "Unindexed".
- Support the new explicit encryption API: `queryType` and `contentionFactor`.

# Background & Motivation

The ticket "Upstream Changes Summary" links to the specification changes implemented.

### CODE REVIEW

File: src/libmongoc/src/mongoc/mongoc-crypt-private.h [hunk 1]
```diff
@@ -87,6 +89,8 @@ _mongoc_crypt_explicit_encrypt (_mongoc_crypt_t *crypt,
                                 const char *algorithm,
                                 const bson_value_t *keyid,
                                 char *keyaltname,
+                                mongoc_encrypt_query_type_t *query_type,
+                                int64_t *contention_factor,
```
Comment (vector-of-bool) [inline src/libmongoc/src/mongoc/mongoc-crypt-private.h:93]: These should be pointers-to-`const`

File: src/libmongoc/src/mongoc/mongoc-client-side-encryption.c [hunk 2]
```diff
@@ -379,6 +379,14 @@ struct _mongoc_client_encryption_encrypt_opts_t {
...
    char *algorithm;
    char *keyaltname;
+   struct {
+      int64_t value;
+      bool set;
+   } contention_factor;
+   struct {
+      mongoc_encrypt_query_type_t value;
+      bool set;
+   } query_type;
```
Comment (vector-of-bool) [inline src/libmongoc/src/mongoc/mongoc-client-side-encryption.c:389]: This idiom is cool and good.

File: src/libmongoc/src/mongoc/mongoc-client-side-encryption.c [hunk 3]
```diff
@@ -438,6 +446,29 @@ mongoc_client_encryption_encrypt_opts_set_algorithm (
...
+void
+mongoc_client_encryption_encrypt_opts_set_contention_factor (
+   mongoc_client_encryption_encrypt_opts_t *opts, int64_t contention_factor)
+{
+   if (!opts) {
+      return;
+   }
+   opts->contention_factor.value = contention_factor;
+   opts->contention_factor.set = true;
+}
```
Comment (vector-of-bool) [inline src/libmongoc/src/mongoc/mongoc-client-side-encryption.c:458]: Do we care about giving the user a way to "un-set" these options?

File: src/libmongoc/src/mongoc/mongoc-client-side-encryption.c (hunk 2 above)
Comment (kevinAlbs) [reply src/libmongoc/src/mongoc/mongoc-client-side-encryption.c:389]: Thank you.

File: src/libmongoc/src/mongoc/mongoc-client-side-encryption.c (hunk 3 above)
Comment (kevinAlbs) [reply src/libmongoc/src/mongoc/mongoc-client-side-encryption.c:458]: IMO no. I do not see much of a use case. A `mongoc_client_encryption_encrypt_opts_reset_contention_factor` could be added later if needed.

File: src/libmongoc/src/mongoc/mongoc-crypt-private.h (hunk 1 above)
Comment (kevinAlbs) [reply src/libmongoc/src/mongoc/mongoc-crypt-private.h]: Done.

**Review by vector-of-bool — APPROVED** [review]

**Pull request merged**